dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/user/UserController.java

```diff
@@ -170,35 +160,92 @@
+
+        return pager;
+    }
 
+    private Query makeQuery( WebOptions options, List<String> filters,
+        List<Order> orders, UserQueryParams params )
+    {
         Pagination pagination = CollectionUtils.isEmpty( filters ) ? new Pagination() : getPaginationData( options );
 
+        List<String> ordersAsString = (orders == null) ? null
+            : orders.stream()
+                .map( Order::toOrderString )
+                .collect( Collectors.toList() );
+
         /*
          * Keep the memory query on the result
          */
-        Query query = queryService.getQueryFromUrl( getEntityClass(),
-            filters, orders, pagination, options.getRootJunction() );
-        query.setDefaultOrder();
-        query.setDefaults( Defaults.valueOf( options.get( "defaults", DEFAULTS ) ) );
+        Query query = queryService
+            .getQueryFromUrl( getEntityClass(), filters, orders, pagination, options.getRootJunction() );
+
+        List<User> users = userService.getUsers( params, ordersAsString );
+
         query.setObjects( users );
+        query.setDefaults( Defaults.valueOf( options.get( "defaults", DEFAULTS ) ) );
+        query.setDefaultOrder();
 
-        return (List<User>) queryService.query( query );
+        return query;
+    }
+
+    private UserQueryParams makeUserQueryParams( WebOptions options )
+    {
+        UserQueryParams params = new UserQueryParams();
+        params.setQuery( StringUtils.trimToNull( options.get( "query" ) ) );
+        params.setPhoneNumber( StringUtils.trimToNull( options.get( "phoneNumber" ) ) );
+        params.setCanManage( options.isTrue( "canManage" ) );
+        params.setAuthSubset( options.isTrue( "authSubset" ) );
+        params.setLastLogin( options.getDate( "lastLogin" ) );
+        params.setInactiveMonths( options.getInt( "inactiveMonths" ) );
+        params.setInactiveSince( options.getDate( "inactiveSince" ) );
+        params.setSelfRegistered( options.isTrue( "selfRegistered" ) );
+        params.setInvitationStatus( UserInvitationStatus.fromValue( options.get( "invitationStatus" ) ) );
+        params.setUserOrgUnits( options.isTrue( "userOrgUnits" ) );
+        params.setIncludeOrgUnitChildren( options.isTrue( "includeChildren" ) );
+
+        return params;
+    }
+
+    /**
+     * Overrides count() method in AbstractController. The reason we override is
+     * because we have more optimized queries for User, since User has some
+     * complicated relationships it can result in less optimized count queries
+     * when using the generic.
+     *
+     * @param options
+     * @param filters
+     * @param orders
+     * @return
+     */
+    protected long count( WebOptions options, List<String> filters, List<Order> orders )
+    {
+        UserQueryParams params = makeUserQueryParams( options );
+        Query query = makeQuery( options, filters, orders, params );
+
+        List<? extends IdentifiableObject> list = queryService.query( query );
```

Comment:
Did we run the same `query` already? 

In that case the `count` method could also get a parameter where the abstract controller passes in the result of the query before it is sliced and limited so that count could simply do: `return results.size()`.

Reply:
Yeah, I agree, but the semantics will be different for the original count. I wanted to make it so that it only affects the user count. The count will be fetched straight from memory though, else I would 100% agree.